### ISSUE TITLE
Misleading error message when evaluating erroneous form in .clj file

### ISSUE THREAD

**Issue by slipset** [clojure-emacs/cider#1372]
Given this form 

```
(identity  {:foo })
```

When evalutating in the cider repl, I get the following error message

```
user> (identity  {:foo })
RuntimeException Map literal must contain an even number of forms  clojure.lang.Util.runtimeException (Util.java:221)
RuntimeException Unmatched delimiter: )  clojure.lang.Util.runtimeException (Util.java:221)
```

which is what is expected.

When evaluating the same form in the .clj buffer with `C-c C-e`, I get this error message

```

  Show: Clojure Java REPL Tooling Duplicates All  (22 frames hidden)

2. Unhandled clojure.lang.LispReader$ReaderException
   java.lang.RuntimeException: Unmatched delimiter: )

               LispReader.java:  292  clojure.lang.LispReader/read
               LispReader.java:  196  clojure.lang.LispReader/read
               LispReader.java:  190  clojure.lang.LispReader/read
                      core.clj: 3640  clojure.core/read
                      core.clj: 3638  clojure.core/read
        interruptible_eval.clj:   68  clojure.tools.nrepl.middleware.interruptible-eval/evaluate/fn/fn
                      main.clj:  238  clojure.main/repl/read-eval-print/fn
                      main.clj:  238  clojure.main/repl/read-eval-print
                      main.clj:  258  clojure.main/repl/fn
                      main.clj:  258  clojure.main/repl
                   RestFn.java: 1523  clojure.lang.RestFn/invoke
        interruptible_eval.clj:   58  clojure.tools.nrepl.middleware.interruptible-eval/evaluate/fn
                      AFn.java:  152  clojure.lang.AFn/applyToHelper
                      AFn.java:  144  clojure.lang.AFn/applyTo
                      core.clj:  630  clojure.core/apply
                      core.clj: 1868  clojure.core/with-bindings*
                   RestFn.java:  425  clojure.lang.RestFn/invoke
        interruptible_eval.clj:   56  clojure.tools.nrepl.middleware.interruptible-eval/evaluate
        interruptible_eval.clj:  188  clojure.tools.nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
        interruptible_eval.clj:  157  clojure.tools.nrepl.middleware.interruptible-eval/run-next/fn
                      AFn.java:   22  clojure.lang.AFn/run
       ThreadPoolExecutor.java: 1142  java.util.concurrent.ThreadPoolExecutor/runWorker
       ThreadPoolExecutor.java:  617  java.util.concurrent.ThreadPoolExecutor$Worker/run
                   Thread.java:  745  java.lang.Thread/run

1. Caused by java.lang.RuntimeException
   Unmatched delimiter: )

                     Util.java:  221  clojure.lang.Util/runtimeException
               LispReader.java: 1176  clojure.lang.LispReader$UnmatchedDelimiterReader/invoke
               LispReader.java:  263  clojure.lang.LispReader/read
               LispReader.java:  196  clojure.lang.LispReader/read
               LispReader.java:  190  clojure.lang.LispReader/read
                      core.clj: 3640  clojure.core/read
                      core.clj: 3638  clojure.core/read
        interruptible_eval.clj:   68  clojure.tools.nrepl.middleware.interruptible-eval/evaluate/fn/fn
                      main.clj:  238  clojure.main/repl/read-eval-print/fn
                      main.clj:  238  clojure.main/repl/read-eval-print
                      main.clj:  258  clojure.main/repl/fn
                      main.clj:  258  clojure.main/repl
                   RestFn.java: 1523  clojure.lang.RestFn/invoke
        interruptible_eval.clj:   58  clojure.tools.nrepl.middleware.interruptible-eval/evaluate/fn
                      AFn.java:  152  clojure.lang.AFn/applyToHelper
                      AFn.java:  144  clojure.lang.AFn/applyTo
                      core.clj:  630  clojure.core/apply
                      core.clj: 1868  clojure.core/with-bindings*
                   RestFn.java:  425  clojure.lang.RestFn/invoke
        interruptible_eval.clj:   56  clojure.tools.nrepl.middleware.interruptible-eval/evaluate
        interruptible_eval.clj:  188  clojure.tools.nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
        interruptible_eval.clj:  157  clojure.tools.nrepl.middleware.interruptible-eval/run-next/fn
                      AFn.java:   22  clojure.lang.AFn/run
       ThreadPoolExecutor.java: 1142  java.util.concurrent.ThreadPoolExecutor/runWorker
       ThreadPoolExecutor.java:  617  java.util.concurrent.ThreadPoolExecutor$Worker/run
                   Thread.java:  745  java.lang.Thread/run

```

which is not the root cause of the error

**Comment by Malabarba** [comment]
Looks like this is the exception being thrown by the reader. So it's not really a cider bug IMO. 

You consider it a clojure bug (reader throwing the wrong exception) or consider it an nrepl bug (if you think nrepl should politely reject invalid code instead of throwing an exception).

**Comment by expez** [comment]
The reader throws the same exception both times, but  when evaluating from the file we swallow half the information...

**Comment by Malabarba** [comment]
> The reader throws the same exception both times, but when evaluating from the file we swallow half the information...

Hmm. On the repl buffer, it looks like two exceptions are being thrown. If so, that explains why the exception buffer only displays one (it simply displays the most recent one).

**Comment by Malabarba** [comment]
I'm more curious about why the repl evaluation doesn't pop up the exception buffer.

**Comment by samroberton** [comment]
A colleague just came across a similar example:

`{:foo :a :foo :b}` results in the expected exception complaining of "Duplicate key", in the `*cider-error*` popup buffer, and with a summary line printed in the REPL buffer.

`(def bar {:foo :a :foo :b})` results in two exceptions being summarised in the REPL buffer:  the expected "Duplicate key", and this "Unmatched delimiter: )".  The *latter* is the one that comes up in the `*cider-error*` buffer, and there doesn't seem to be any way to get the former, which is the actual root cause.

This affects loading from a file, too.

A bare `lein repl` (not in Emacs) also prints the two exceptions, too, and binds `*e` to the latter (the unhelpful one).

The root cause seems to be this issue:  http://dev.clojure.org/jira/browse/CLJ-1146

**Comment by alexander-yakushev** [comment]
Bump. Anything we can do about this? The commenter in the JIRA ticket claims that the issue cannot be easily solved on the Clojure side. All of the REPLs I've tried (`lein repl`, `boot repl`) throw double exceptions when the encounter a nested form with odd/duplicate elements in `{}`.

This is one of the very few bugs that can be solved only by *long and concentrated stare* at the code. Neither Clojure nor CIDER gives any idea what actually went wrong. I (and probably others) have learned that the suspicious double-blip of the exception window is likely attributed to the map reader, but this is a tribal knowledge — it is really frustrating for the beginners.

**Comment by xiongtx** [comment]
I investigated this a bit and posted some thoughts on [CLJ-1146](https://dev.clojure.org/jira/browse/CLJ-1146).

The problem is in the [`loop`](https://github.com/clojure/clojure/blob/clojure-1.9.0/src/clj/clojure/main.clj#L259-L269) of `clojure.main/repl`. When the reader throws an exception, the exception is processed by `caught` (default [`repl-caught`](https://github.com/clojure/clojure/blob/clojure-1.9.0/src/clj/clojure/main.clj#L152), which prints a short ST), then `*e` is set to the exception. _However_, the loop continues with the remaining chars, which causes `read` to throws _another_ exception, and so on.

The solution would be to flush the input in `caught` so the REPL prompts for a new input, leaving `*e` as the "real" error.

This change would ideally be made in `clojure.main/repl`, but in light of the scant attention given to `CLJ-1146` I'm not optimistic. Perhaps the change could be [made in `tools.nrepl`](https://github.com/clojure/tools.nrepl/blob/tools.nrepl-0.2.13/src/main/clojure/clojure/tools/nrepl/middleware/interruptible_eval.clj#L117-L125) instead? cc @cemerick. If it can be made in `cider-nrepl`, someone'll have to point me to the spot--I can't find it.

I think this change would go a long way towards making Clojure's errors more tractable. People complain about stacktraces, but it's not really STs themselves that are the problem--it's STs that have nothing to do with the real issue. If we can report `Map literal must contain an even number of forms` instead of `Unmatched delimiter: )`, that'd be a huge win :trophy:.

**Comment by alexander-yakushev** [comment]
This should be fixed by https://github.com/nrepl/nrepl/pull/107 which is released as nrepl 0.6.0 and is already available in Leiningen 2.9.0 🎉 .
